### PR TITLE
Geomap: configure legend on map

### PR DESCRIPTION
**Pull request by lean257**
**What this PR does / why we need it**:
This PR adds the legend feature. By default, legend will show up when loading geomap. 

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/plugins/panel/geomap/layers/data/MarkersLegend.tsx [hunk 1]
```diff
@@ -1,44 +1,49 @@
...
+      return <></>;
+    }
+
+    if (!color.field && color.fixed) {
+      return <div className={this.style.infoWrap}>Fixed: {color.fixed}</div>;
+    }
+
+    const colorMode = getFieldColorModeForField(color.field!);
+    if (colorMode.isContinuous && colorMode.getColors) {
+      return <div className={this.style.infoWrap}>Color scale</div>;
```
Comment (ryantxu) [inline public/app/plugins/panel/geomap/layers/data/MarkersLegend.tsx:36]: @lean257 -- can you look at implementing this variation?  I was unable to get things to update predictably :frowning_face: it looks like you need to save and full refresh to get here

**Comment by ryantxu** [comment]
I just pushed a commit that shows:
1. scale min max
2. field: min/max mean

image

Ideally the continuous legend element could somehow incorporate all these

**Comment by ryantxu** [comment]
Also moved the "showLegend" setting form the 'Map controls section' into the markers layer.  This will let us have more control over what we actually want to do in the legend (maybe interactive, etc)

image

Note, existing panels will have an odd first load since the showLegend value is not present and defaults to true

**Comment by lean257** [comment]
I'm currently using the values set by users to set the legend min/max, so once they change the options it will reflect in the legend as well. 
Screen Shot 2021-07-26 at 3 14 38 PM

File: public/app/plugins/panel/geomap/layers/data/MarkersLegend.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    }
+  `,
+  legendItem: css`
+    white-space: nowrap;
+  `,
+  gradientContainer: css`
+    min-width: 200px;
+    display: flex;
+    justify-content: space-between;
+    background-image: linear-gradient(to right, ${colors[0]}, ${colors[1]}, ${colors[2]});
```
Comment (ryantxu) [inline public/app/plugins/panel/geomap/layers/data/MarkersLegend.tsx:115]:
```suggestion
    background-image: linear-gradient(to right, ${colors.map( c => ... });
```
Do we know the color array length?

Comment (lean257) [reply public/app/plugins/panel/geomap/layers/data/MarkersLegend.tsx:115]: yeah length in this case is 3 but maybe not always.

File: public/app/plugins/panel/geomap/layers/data/markersLayer.tsx [hunk 3]
```diff
@@ -11,6 +11,11 @@ import { getScaledDimension, } from '../../dims/scale';
 import { getColorDimension, } from '../../dims/color';
 import { ScaleDimensionEditor } from '../../dims/editors/ScaleDimensionEditor';
 import { ColorDimensionEditor } from '../../dims/editors/ColorDimensionEditor';
+import React from 'react';
```
Comment (ryantxu) [inline public/app/plugins/panel/geomap/layers/data/markersLayer.tsx:14]:
```suggestion
import React { ReactNode } from 'react';
```
usually we have react as the first import -- not sure why :man_shrugging:

**Review by ryantxu — APPROVED** [review]
:tada:

File: public/app/plugins/panel/geomap/layers/data/markersLayer.tsx [hunk 4]
```diff
@@ -107,8 +125,17 @@ export const markersLayer: MapLayerRegistryItem<MarkersConfig> = {
...
 
+          // Post updates to the legend component
+          if (legend) {
+            console.log( 'UPDATE (marker layer)', colorDim);
+            legendProps.next({
+              color: colorDim,
+              size: sizeDim,
+            });
+          }
+          break; // Only the first frame for now!
```
Comment (nicolaerosia) [inline public/app/plugins/panel/geomap/layers/data/markersLayer.tsx:137]: why only the first frame? Is this bug related, https://github.com/grafana/grafana/issues/48274 ? (credits to @yoyo421)

**Pull request merged**